insights/core/__init__.py

```diff
@@ -229,6 +239,124 @@ def get(self, item, default=None):
         return self.data.get(item, default)
 
 
+class XMLParser(LegacyItemAccess, Parser):
+    """
+    A parser class that reads XML files.  Base your own parser on this.
+    Examples:
+        >>> content = '''
+        ... <?xml version="1.0"?>
+        ... <data xmlns:fictional="http://characters.example.com"
+        ...       xmlns="http://people.example.com">
+        ...     <country name="Liechtenstein">
+        ...         <rank updated="yes">2</rank>
+        ...         <year>2008</year>
+        ...         <gdppc>141100</gdppc>
+        ...         <neighbor name="Austria" direction="E"/>
+        ...         <neighbor name="Switzerland" direction="W"/>
+        ...     </country>
+        ...     <country name="Singapore">
+        ...         <rank updated="yes">5</rank>
+        ...         <year>2011</year>
+        ...         <gdppc>59900</gdppc>
+        ...         <neighbor name="Malaysia" direction="N"/>
+        ...     </country>
+        ...     <country name="Panama">
+        ...         <rank>68</rank>
+        ...         <year>2011</year>
+        ...         <gdppc>13600</gdppc>
+        ...         <neighbor name="Costa Rica" direction="W"/>
+        ...     </country>
+        ... </data>
+        ... '''.strip()
+        >>> xml_parser = XMLParser(context_wrap(content))
+        >>> xml_parser.xmlns
+        'http://people.example.com'
+        >>> xml_parser.get_elements(".")[0].tag # Top-level elements
+        'data'
+        >>> len(xml_parser.get_elements("./country/neighbor", None)) # All 'neighbor' grand-children of 'country' children of the top-level elements
+        3
+        >>> len(xml_parser.get_elements(".//year/..[@name='Singapore']")[0]) # Nodes with name='Singapore' that have a 'year' child
+        1
+        >>> xml_parser.get_elements(".//*[@name='Singapore']/year")[0].text # 'year' nodes that are children of nodes with name='Singapore'
+        '2011'
+        >>> xml_parser.get_elements(".//neighbor[2]", "http://people.example.com")[0].get('name') # All 'neighbor' nodes that are the second child of their parent
+        'Switzerland'
+    Attributes:
+        raw (str): raw XML content
+        dom (Element): Root element of parsed XML file
+        xmlns (str): The default XML namespace, an empty string when no
+            namespace is declared.
+        data (dict): All required specific properties can be included in data.
+    """
+
+    def _parse_dom(self):
```

Comment:
@zhangyi733 see my comment in #856, since this can be implemented in a child class to change functionality it should either be a public method so that it will show up in the documentation or at least described in the class docstring so that users will be aware of its functionality.  Either way, the documentation should be clear as to how it is used.